docs/examples/201/vm-new-or-existing-conditions/main.bicep

```diff
@@ -0,0 +1,263 @@
+}
+param vmName string {
+  metadata: {
+    description: 'Name for the Virtual Machine.'
+  }
+  default: 'linux-vm'
+}
+param adminUsername string {
+  metadata: {
+    description: 'User name for the Virtual Machine.'
+  }
+}
+param authenticationType string {
+  allowed: [
+    'password'
+    'sshPublicKey'
+  ]
+  metadata: {
+    description: 'Type of authentication to use on the Virtual Machine.'
+  }
+  default: 'sshPublicKey'
+}
+param adminPasswordOrKey string {
+  metadata: {
+    description: 'Password or ssh key for the Virtual Machine.'
+  }
+  secure: true
+}
+param vmSize string {
+  metadata: {
+    description: 'Size for the Virtual Machine.'
+  }
+  default: 'Standard_A2_v2'
+}
+param createNewStorageAccount bool {
+  metadata: {
+    description: 'Determines whether or not a new storage account should be provisioned.'
+  }
+  default: true
+}
+param storageAccountName string {
+  metadata: {
+    description: 'Name of the storage account'
+  }
+  default: 'storage${uniqueString(resourceGroup().id)}'
+}
+param storageAccountType string {
+  metadata: {
+    description: 'Storage account type'
+  }
+  default: 'Standard_LRS'
+}
+param storageAccountResourceGroupName string {
+  metadata: {
+    description: 'Name of the resource group for the existing storage account'
+  }
+  default: resourceGroup().name
+}
+param createNewVnet bool {
+  metadata: {
+    description: 'Determines whether or not a new virtual network should be provisioned.'
+  }
+  default: true
+}
+param vnetName string {
+  metadata: {
+    description: 'Name of the virtual network'
+  }
+  default: 'VirtualNetwork'
+}
+param addressPrefixes array {
+  metadata: {
+    description: 'Address prefix of the virtual network'
+  }
+  default: [
+    '10.0.0.0/16'
+  ]
+}
+param subnetName string {
+  metadata: {
+    description: 'Name of the subnet'
+  }
+  default: 'default'
+}
+param subnetPrefix string {
+  metadata: {
+    description: 'Subnet prefix of the virtual network'
+  }
+  default: '10.0.0.0/24'
+}
+param vnetResourceGroupName string {
+  metadata: {
+    description: 'Name of the resource group for the existing virtual network'
+  }
+  default: resourceGroup().name
+}
+param createNewPublicIP bool {
+  metadata: {
+    description: 'Determines whether or not a new public ip should be provisioned.'
+  }
+  default: true
+}
+param publicIPName string {
+  metadata: {
+    description: 'Name of the public ip address'
+  }
+  default: 'PublicIp'
+}
+param publicIPDns string {
+  metadata: {
+    description: 'DNS of the public ip address for the VM'
+  }
+  default: 'linux-vm-${uniqueString(resourceGroup().id)}'
+}
+param publicIPResourceGroupName string {
+  metadata: {
+    description: 'Name of the resource group for the public ip address'
+  }
+  default: resourceGroup().name
+}
+
+var storageAccountId = createNewStorageAccount ? storageAccount.id : resourceId(storageAccountResourceGroupName, 'Microsoft.Storage/storageAccounts/', storageAccountName)
```

Comment:
so this pattern is the weird part right? because in the template today you would use the `resourceId()` function for both new and existing and because the storageAccountName is the same in both cases it happens to work. With bicep we have arguably made it more complex (though I'd argue more obvious).

Going forward this will be something like:

```
external resource existingStorage ...

var storageAccountId = createNewStorageAccount ? storageAccount.id :  existingStorageAccount.id
```

Which seems good to me

Reply:
oh wait nevermind, I got that wrong. It will be:

```
resource storageAccount 'Microsoft.Storage/storageAccounts@2017-06-01' = if (createNewStorageAccount) {
  name: storageAccountName
  location: location
  kind: 'Storage'
  sku: {
    name: storageAccountType
  }
} else existing ...
```

Then you use the same symbolic name regardless if the storage account is being created or not. So I think we are good.